teos/cli/teos_cli.py

```diff
@@ -1,194 +1,328 @@
-        return to_json(list(result.user_ids))
+        teos_rpc_host = config.get("RPC_BIND")
+        teos_rpc_port = config.get("RPC_PORT")
 
-    @formatted
-    def get_user(self, user_id):
-        """
-        Gets information about a specific user.
+        self.rpc_client = RPCClient(teos_rpc_host, teos_rpc_port)
 
-        Args:
-            user_id (:obj:`str`): the id of the requested user.
+    @classmethod
+    def command(cls, command_cls):
+        """
+        Decorator used to register a new command, which must be a subclass of :class:`CLICommand` and must override
+        the ``name`` field with an appropriate string.
 
         Raises:
-            :obj:`InvalidParameter`: if `user_id` is not in the valid format.
+            :obj:`TypeError`: ``command_cls`` is not a subclass of :class:`CLICommand`, or its ``name`` field is not a
+            string.
         """
 
-        if not is_compressed_pk(user_id):
-            raise InvalidParameter("Invalid user id")
+        if not issubclass(command_cls, CLICommand):
+            raise TypeError(f"{command_cls.__name__} is not a subclass of CLICommand")
+
+        if not isinstance(command_cls.name, str):
+            raise TypeError(f'The "name" attribute of {command_cls.__name__} must be a string.')
+
+        cls.COMMANDS[command_cls.name] = command_cls
+        return command_cls
 
-        result = self.stub.get_user(GetUserRequest(user_id=user_id))
-        return result.user
+    def run(self, command_name, raw_args):
+        """
+        Parses ``raw_args`` using the ``parse_args`` method of the command.
+        Then, executes the command's ``run`` method, passing the ``rpc_client`` and the output of ``parse_args`` to it.
+        If any error is raised by the command, returns an error message.
 
-    def stop(self):
-        """Stops TEOS gracefully."""
-        self.stub.stop(Empty())
-        print("Closing the Eye of Satoshi")
+        Returns:
+            :obj:`str`: The return value of the ``run`` method of the command or an error message.
+        """
 
+        if command_name not in self.COMMANDS:
+            return "Unknown command. Use help to check the list of available commands"
 
-def main(command, args, data_dir, command_line_conf):
-    # Loads config and sets up the data folder and log file
-    config_loader = ConfigLoader(data_dir, CONF_FILE_NAME, DEFAULT_CONF, command_line_conf)
-    config = config_loader.build_config()
+        command = self.COMMANDS[command_name]
 
-    setup_data_folder(data_dir)
+        try:
+            args = command.parse_args(raw_args)
+            return command.run(self.rpc_client, args)
+        except grpc.RpcError as e:
+            if e.code() == grpc.StatusCode.UNAVAILABLE:
+                return "It was not possible to reach the Eye of Satoshi. Are you sure the tower is running?"
+            else:
+                return e.details()
+        except InvalidParameter as e:
```

Comment:
I'd remove this `InvalidParameter` catch and let it be captured by the outer scope.

Reply:
Actually, I'm not sure this fits the current design. CLI.run's documentation says that `If any error is raised by the command, returns an error message.`, so it wouldn't make sense to have an exception for InvalidParameter, imho.

If anything, the error handling in in the main `run()` method is redundant if we keep this design.
Rather, I think the CLI.run() method should return the concatenation of the error message and the result of show_usage(), if we want to show the usage for InvalidParameter.

Reply:
That's fine, if you rather sanitise in `.run` and return the user feedback to `main` I'm OK with it.

Reply:
Done in 6b8b5975d4c2884280767a97117c7f181ce2041e (which is rebased to merge with the previous commit).